monai/utils/factories.py

```diff
@@ -0,0 +1,151 @@
+of the factory so that they can function as constant identifiers. eg. instance normalization is named `Norm.INSTANCE`.
+
+For example, to get a transpose convolution object the name is needed and then a dimension argument is provided which is
+passed to the factory function:
+
+.. code-block:: python
+
+    dimension = 3
+    name = Conv.CONVTRANS
+    conv = Conv[name, dimension]
+
+This allows the `dimension` value to be set in the constructor, for example so that the dimensionality of a network is
+parameterizable. Not all factories require arguments after the name, the caller must be aware which are required.
+
+Defining new factories involves creating the object then associating it with factory functions:
+
+.. code-block:: python
+
+    fact = ObjectFactory()
+
+    @fact.factory_function('test')
+    def make_something(x, y):
+        # do something with x and y to choose which layer type to return
+        return SomeLayerType
+    ...
+
+    # request object from factory TEST with 1 and 2 as values for x and y
+    layer = fact[fact.TEST, 1, 2]
+
+Typically the caller of a factory would know what arguments to pass (ie. the dimensionality of the requested type) but
+can be parameterized with the factory name and the arguments to pass to the created type at instantiation time:
+
+.. code-block:: python
+
+    def use_factory(fact_args):
+        fact_name, type_args = split_args
+        layer_type = fact[fact_name, 1, 2]
+        return layer_type(**type_args)
+    ...
+
+    kw_args = {'arg0':0, 'arg1':True}
+    layer = use_factory( (fact.TEST, kwargs) )
+"""
+
+from typing import Any, Callable, Dict, Tuple
+
+__all__ = ["ObjectFactory"]
+
+
+class ObjectFactory:
+    """
+    Factory object for creating objects, this uses given factory functions to actually produce the types or constructing
+    callables. These functions are referred to by name and can be added at any time.
+    """
+
+    def __init__(self) -> None:
+        self.factories: Dict[str, Callable] = {}
+
+    @property
+    def names(self) -> Tuple[str, ...]:
+        """
+        Produces all factory names.
+        """
+
+        return tuple(self.factories)
+
+    def add_factory_callable(self, name: str, func: Callable) -> None:
+        """
+        Add the factory function to this object under the given name.
+        """
+
+        self.factories[name.upper()] = func
+        self.__doc__ = (
+            "The supported member"
+            + ("s are: " if len(self.names) > 1 else " is: ")
+            + ", ".join(f"``{name}``" for name in self.names)
+            + ".\nPlease see :py:class:`monai.networks.layers.split_args` for additional args parsing."
+        )
+
+    def factory_function(self, name: str) -> Callable:
+        """
+        Decorator for adding a factory function with the given name.
+        """
+
+        def _add(func: Callable) -> Callable:
+            self.add_factory_callable(name, func)
+            return func
+
+        return _add
```

Comment:
I'd suggest this decorator method for class types:

```python
def factory_type(self, name: Optional[str]=None) -> Callable:
    """
    Decorator for class types, adding a factory function taking no arguments. This is stored under `name` or the class'
    name if not provided.
    """
    def _add(cls: type) -> type:
        self.add_factory_callable(name or cls.__name__, lambda: cls)
        return cls

    return _add
```

This would be used as 

```python
fact=ObjectFactory()

@fact.factory_type()
class Foo:
    def __init__(self,a,b):
        pass

foo_inst = fact["Foo"](a=1,b=2)
```

I *think* that'll work but I haven't tried it.